semcore/popper/src/modifiers/cursorAnchoring.ts

```diff
@@ -0,0 +1,55 @@
+import getOffsetParent from '@popperjs/core/lib/dom-utils/getOffsetParent';
+import { Modifier } from '@popperjs/core';
+
+const clamp = (value: number, boundary1: number, boundary2: number) => {
+  const min = Math.min(boundary1, boundary2);
+  const max = Math.max(boundary1, boundary2);
+  return Math.min(Math.max(value, min), max);
+};
+
+type Options = {
+  cursorAnchoring?: 'horizontal' | 'vertical';
+  mouseEnterCursorPositionRef: { current: { x: number; y: number } };
+};
+
+const getWindowRect = () => ({ x: 0, y: 0, width: window.innerWidth, height: window.innerHeight });
+
+export const cursorAnchoringModifier: Modifier<'cursorAnchoring', Options> = {
+  name: 'cursorAnchoring',
+  phase: 'beforeRead',
+  enabled: true,
+  fn: ({ state, options }) => {
+    const { cursorAnchoring, mouseEnterCursorPositionRef } = options;
+    const { rects, elements } = state;
+    if (cursorAnchoring === 'horizontal') {
+      const offsetParent = getOffsetParent(elements.popper);
+      const offsetParentRect =
+        offsetParent === window ? getWindowRect() : offsetParent.getBoundingClientRect();
+      const mouseX = mouseEnterCursorPositionRef?.current?.x;
+      if (mouseX === undefined) return;
+      const width = Math.min(rects.reference.width, rects.popper.width);
+      const x = clamp(
+        mouseX - offsetParentRect.x - width / 2,
+        rects.reference.x,
+        rects.reference.x + rects.reference.width - width,
+      );
```

Comment:
let's talk about this math's... maybe we should add some comments

Reply:
Not sure that comments will make it more clear but I can ensure you that it's correct